src/main/java/com/redhat/rhjmc/containerjfr/net/TargetConnectionManager.java

```diff
@@ -0,0 +1,149 @@
+        this.logger = logger;
+        this.jfrConnectionToolkit = jfrConnectionToolkit;
+    }
+
+    public <T> T executeConnectedTask(String targetId, ConnectedTask<T> task) throws Exception {
+        try {
+            if (activeConnections.containsKey(targetId)) {
+                return task.execute(activeConnections.get(targetId));
+            } else {
+                try (JFRConnection connection = connect(targetId)) {
+                    activeConnections.put(targetId, connection);
+                    return task.execute(connection);
+                }
+            }
+        } finally {
+            activeConnections.remove(targetId);
+        }
+    }
+
+    /**
+     * Returns a new JFRConnection to the specified Target. This does not do any connection reuse or
+     * other management, so clients are responsible for cleaning up the connection when they are
+     * finished with it. When possible, clients should use executeConnectedTask instead, which does
+     * perform automatic cleanup when the provided task has been completed.
+     */
+    public JFRConnection connect(String targetId) throws Exception {
+        try {
+            return attemptConnectAsJMXServiceURL(targetId);
+        } catch (Exception e) {
+            return attemptConnectAsHostPortPair(targetId);
+        }
+    }
+
+    private JFRConnection attemptConnectAsJMXServiceURL(String url) throws Exception {
+        return connect(new JMXServiceURL(url));
+    }
+
+    private JFRConnection attemptConnectAsHostPortPair(String s) throws Exception {
+        Matcher m = HOST_PORT_PAIR_PATTERN.matcher(s);
+        if (!m.find()) {
+            return null;
+        }
+        String host = m.group(1);
+        String port = m.group(2);
+        if (port == null) {
+            port = "9091";
+        }
+        return connect(host, Integer.parseInt(port));
+    }
+
+    private JFRConnection connect(JMXServiceURL url) throws Exception {
+        logger.trace(String.format("Locking connection %s", url.toString()));
+        lock.lockInterruptibly();
+        return jfrConnectionToolkit.connect(
+                url,
+                List.of(
+                        lock::unlock,
+                        () ->
+                                logger.trace(
+                                        String.format("Unlocking connection %s", url.toString()))));
+    }
+
+    private JFRConnection connect(String host, int port) throws Exception {
+        logger.trace(String.format("Locking connection %s:%d", host, port));
+        lock.lockInterruptibly();
+        return jfrConnectionToolkit.connect(
+                host,
+                port,
+                List.of(
+                        lock::unlock,
+                        () ->
+                                logger.trace(
+                                        String.format("Unlocking connection %s:%d", host, port))));
```

Comment:
Yes exactly, it's a list of functions to be invoked when the underlying connection is closed. I already went ahead and pushed these additions in container-jfr-core (v0.12.x), since they were relatively trivial - the actual connection close detection and cleanup handling is done by the JMC classes we're using behind the scenes, but -core used to just pass in an empty list of cleanup functions every time.

Reply:
Oh cool, thanks for the explanation. I was looking in master for the changes, I see them now in 0.12.3.